### PR TITLE
String to Generic Type Parser

### PR DESCRIPTION
**Pull request by alanedwardes**
Adds `.Parse<T>` and `.TryParse<T>` string extensions, which can parse the following types:

- Enum from string
- Enum decorated with `EnumMember` attributes, so each enum member can have a different value
- int, bool, long, float, double, Guid

I have done some basic perf optimisation, this is around 2x slower than the strongly typed .Parse/.TryParse methods in most cases, which imo is a good tradeoff. `ParsingBenchmarks` has unit tests comparing the built in typed methods to this generic parser over 10 million iterations.

### CODE REVIEW

File: test/Narochno.Primitives.Parsing.Tests/project.json [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+  "testRunner": "xunit",
+
+  "dependencies": {
+    "dotnet-test-xunit": "2.2.0-preview2-build1029",
+    "Narochno.Primitives.Parsing": { "target": "project" },
+    "xunit": "2.2.0-beta4-build3444"
+  },
+
+  "frameworks": {
+    "netcoreapp1.0": {
```
Comment (adamhathcock) [inline test/Narochno.Primitives.Parsing.Tests/project.json]: 1.1?

**Comment by adamhathcock** [comment]
Maybe look at this for benchmarking?  https://petabridge.com/blog/introduction-to-nbench/

**Comment by alanedwardes** [comment]
@adamhathcock Separate library because it pulls in `System.Reflection` and `System.Reflection.TypeExtensions`, I didn't want those as dependencies for the main primitives package

**Comment by alanedwardes** [comment]
@adamhathcock NBench looks like it's net451 only, don't really want to invest the time. Can be for a separate PR anyway

**Comment by adamhathcock** [comment]
I feel system libraries should be okay to pull in.  But Json is rightfully a separate package.  Just balancing the number of packages.  If you feel strong then it's cool.

I didn't see that about NBench.  Just knew of it as a library and thought it might be useful.  Guess it's not.

**Comment by alanedwardes** [comment]
Yeah I don't feel strongly, leaning towards combining now to cut down number of packages... going to try it

**Comment by alanedwardes** [comment]
I think it should be separate, the packages it needs are:

```
    "System.Collections": "4.3.0",
    "System.Reflection": "4.3.0",
    "System.Reflection.Extensions": "4.3.0",
    "System.Reflection.TypeExtensions": "4.3.0",
    "System.Runtime.Serialization.Primitives": "4.3.0"
```

Still on the fence about whether it's worth making a custom attribute instead of using the `EnumMemberAttribute` from `System.Runtime.Serialization.Primitives` to remove that dependency too... I like using something that exists but it's designed for a slightly different purpose

**Comment by adamhathcock** [comment]
I've seen that but don't know how it's used.  Don't want to overload it.  Can always change.

I still think the system packages are okay to share.  If you added a net451 target then they would be just in the framework.

**Comment by alanedwardes** [comment]
Updated

**Comment by alanedwardes** [comment]
Finished with this, will do any target changes in a separate PR, this one is too big

**Pull request merged**